File: ribosome/src/template.ml
module Input = Template_input
module Select = Template_select
module Button = Template_button
module Text = Template_text
module Image = Template_image
module Badge = Template_badge
module Stat = Template_stat
module Divider = Template_divider
module Container = Template_container
module List = Template_list
module Submittable = Template_submittable
module Tone = Template_tone
module Diagram = Template_diagram
module Code = Template_code
module Definition = Template_definition
module Registry = Template_registry
module Codec_error = Codec_error
module Decode = Codec_decode
module Encode = Codec_encode

type t =
  | Text of Template_text.t
  | Image of Template_image.t
  | Badge of Template_badge.t
  | Stat of Template_stat.t
  | Divider of Template_divider.t
  | Diagram of Template_diagram.t
  | Code of Template_code.t
  | Container of t Template_container.t
  | List of t Template_list.t
  | Submittable of Template_submittable.t

let id = function
  | Text t -> t.id
  | Image t -> t.id
  | Badge t -> t.id
  | Stat t -> t.id
  | Divider t -> t.id
  | Diagram t -> t.id
  | Code t -> t.id
  | Container t -> t.id
  | List t -> t.id
  | Submittable t -> t.id

let children = function
  | Container c -> Some c.children
  | List l -> Some l.children
  | _ -> None

let rec decode json =
  let open Codec_decode in
  let* kind = field "kind" string json in
  match kind with
  | "text" ->
      let* t = Text.decode json in
      Ok (Text t)
  | "image" ->
      let* t = Image.decode json in
      Ok (Image t)
  | "badge" ->
      let* t = Badge.decode json in
      Ok (Badge t)
  | "stat" ->
      let* t = Stat.decode json in
      Ok (Stat t)
  | "divider" ->
      let* t = Divider.decode json in
      Ok (Divider t)
  | "diagram" ->
      let* t = Diagram.decode json in
      Ok (Diagram t)
  | "code" ->
      let* t = Code.decode json in
      Ok (Code t)
  | "container" ->
      let* t = Container.decode_child decode json in
      Ok (Container t)
  | "list" ->
      let* t = List.decode decode json in
      Ok (List t)
  | "submittable" ->
      let* t = Submittable.decode json in
      Ok (Submittable t)
  | "input" | "select" | "button" ->
      Error
        (Codec_error.make [ Field "kind" ] InvalidValue
           ("kind '" ^ kind ^ "' is nested-only and cannot appear at root"))
  | _ ->
      Error
        (Codec_error.make [ Field "kind" ] UnknownEnum ("unknown kind: " ^ kind))

let rec encode = function
  | Text t -> Text.encode t
  | Image t -> Image.encode t
  | Badge t -> Badge.encode t
  | Stat t -> Stat.encode t
  | Divider t -> Divider.encode t
  | Diagram t -> Diagram.encode t
  | Code t -> Code.encode t
  | Container t -> Container.encode_child encode t
  | List t -> List.encode encode t
  | Submittable t -> Submittable.encode t

let decode_string s =
  match Yojson.Safe.from_string s with
  | json -> decode json
  | exception _ -> Error (Codec_error.make [] WrongType "invalid JSON string")

let encode_string t = Yojson.Safe.to_string (encode t)

File: frontends/packages/ui-core/src/types/template.ts
export type Tone = "Default" | "Positive" | "Negative" | "Warning" | "Info";

export type Position = { x: number; y: number };
export type Size = { width: number; height: number };

export type DiagramPrimitive =
  | { kind: "text"; text: string; position: Position; tone: Tone }
  | { kind: "line"; start: Position; stop: Position; tone: Tone }
  | { kind: "arrow"; start: Position; stop: Position; tone: Tone }
  | { kind: "rectangle"; origin: Position; size: Size; tone: Tone }
  | { kind: "circle"; center: Position; radius: number; tone: Tone }
  | { kind: "polyline"; points: Position[]; tone: Tone };

export type CodeHighlight = {
  start_line: number;
  end_line: number;
  tone: Tone;
};

export type SelectOption = {
  value: string;
  label: string;
};

export type InputValue = string | number;

export type TextType = "Paragraph" | "H1" | "H2" | "H3" | "H4" | "H5" | "H6";

export type BadgeVariant = "Neutral" | "Success" | "Warning" | "Error" | "Info";

export type ContainerDirection = "Vertical" | "Horizontal";

export type ButtonAction = string;

export type TemplateInput = {
  kind: "input";
  id: string;
  value?: InputValue;
};

export type TemplateSelect = {
  kind: "select";
  id: string;
  label: string;
  options: SelectOption[];
  selected?: string;
};

export type TemplateButton = {
  kind: "button";
  id: string;
  label: string;
  action: ButtonAction;
  disabled?: boolean;
};

export type TemplateText = {
  kind: "text";
  id: string;
  text_type: TextType;
  value: string;
};

export type TemplateImage = {
  kind: "image";
  id: string;
  src: string;
  alt: string;
};

export type TemplateBadge = {
  kind: "badge";
  id: string;
  label: string;
  variant: BadgeVariant;
};

export type TemplateStat = {
  kind: "stat";
  id: string;
  label: string;
  value: string;
  secondary?: string;
};

export type TemplateDivider = {
  kind: "divider";
  id: string;
  label?: string;
};

export type TemplateDiagram = {
  kind: "diagram";
  id: string;
  title: string;
  size: Size;
  primitives: DiagramPrimitive[];
};

export type TemplateCode = {
  kind: "code";
  id: string;
  path: string;
  language: string;
  line_start: number;
  source: string;
  highlights: CodeHighlight[];
};

export type TemplateContainer = {
  kind: "container";
  id: string;
  direction: ContainerDirection;
  children: Template[];
};

export type TemplateList = {
  kind: "list";
  id: string;
  children: Template[];
  ordered?: boolean;
};

export type TemplateSubmittable = {
  kind: "submittable";
  id: string;
  value: (TemplateInput | TemplateSelect)[];
  button?: TemplateButton;
};

export type Template =
  | TemplateText
  | TemplateImage
  | TemplateBadge
  | TemplateStat
  | TemplateDivider
  | TemplateDiagram
  | TemplateCode
  | TemplateContainer
  | TemplateList
  | TemplateSubmittable;

export type NestedTemplate = TemplateInput | TemplateSelect | TemplateButton;

export const TOP_LEVEL_KINDS = [
  "text",
  "image",
  "badge",
  "stat",
  "divider",
  "diagram",
  "code",
  "container",
  "list",
  "submittable",
] as const;

export const NESTED_ONLY_KINDS = ["input", "select", "button"] as const;

export const ALL_KINDS = [...TOP_LEVEL_KINDS, ...NESTED_ONLY_KINDS] as const;

export const TONE_VALUES = ["Default", "Positive", "Negative", "Warning", "Info"] as const;

export const TEXT_TYPE_VALUES = ["Paragraph", "H1", "H2", "H3", "H4", "H5", "H6"] as const;

export const BADGE_VARIANT_VALUES = ["Neutral", "Success", "Warning", "Error", "Info"] as const;

export const CONTAINER_DIRECTION_VALUES = ["Vertical", "Horizontal"] as const;

export type TemplateKind = (typeof ALL_KINDS)[number];
